### PR TITLE
[Chat] Sync API

### PR DESCRIPTION
**Pull request by flypaper0**
# Description
- Sync API client + integration tests
- SyncStore and store delegates for Chat
- Chat SyncAPI integrations

SyncStores: 
- SentInvite including (inviterPubKeyY, inviterPrivKeyY, responseTopic, symKey)
- InviteKey including (pubKey, privKey)
- Thread including (symKey)

* [x] Breaking change
* [x] Requires a documentation update

### CODE REVIEW

File: Sources/WalletConnectSync/Stores/SyncObjectStore.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        var map = getMap(topic: topic)
+        map[id] = nil
+        store.set(element: map, for: topic)
+        return true
+    }
+}
+
+private extension SyncObjectStore {
+
+    func isExists(id: String, topic: String) -> Bool {
```
Comment (llbartekll) [inline Sources/WalletConnectSync/Stores/SyncObjectStore.swift:56]: I am not a language expert but don't like the func name `isExists`

File: Sources/WalletConnectSync/Stores/SyncStore.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        }
+    }
+}
+
+private extension SyncStore {
+
+    func setupSubscriptions() {
+        syncClient.updatePublisher.sink { [unowned self] (topic, update) in
+
+            let record = try! indexStore.getRecord(topic: topic)
```
Comment (llbartekll) [inline Sources/WalletConnectSync/Stores/SyncStore.swift:89]: is it safe to force unwrap?
if so, can you add a comment why?

File: Sources/WalletConnectUtils/NewKeyedDatabase.swift [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import Foundation
+
+public class NewKeyedDatabase<Element> where Element: Codable & Equatable {
```
Comment (llbartekll) [inline Sources/WalletConnectUtils/NewKeyedDatabase.swift:3]: not sure what's the difference but what will be the name of the next one NewNewKeyedDatabase? xd
maybe KeyedDatabaseV2?

**Review by llbartekll — COMMENTED** [review]
Great work,
I would just consider the interface segregation in crypto provider

File: Sources/WalletConnectSigner/CryptoProvider.swift [hunk 4]
```diff
@@ -1,6 +1,12 @@
...
 
+public enum DerivationPath {
+    case hardened(UInt32)
+    case notHardened(UInt32)
+}
+
 public protocol CryptoProvider {
     func recoverPubKey(signature: EthereumSignature, message: Data) throws -> Data
     func keccak256(_ data: Data) -> Data
+    func derive(entropy: Data, path: [DerivationPath]) -> Data
```
[the remark this answers is not in the document]
Comment (flypaper0) [reply Sources/WalletConnectSigner/CryptoProvider.swift]: created DerivationProvider

File: Example/Showcase/Classes/DomainLayer/Chat/ChatService.swift [hunk 5]
```diff
@@ -8,7 +8,7 @@ typealias Stream<T> = AnyPublisher<T, Never>
 final class ChatService {
 
     private var client: ChatClient = {
-        Chat.configure(crypto: DefaultCryptoProvider())
+        Chat.configure(derivator: DefaultDerivationProvider())
```
Comment (llbartekll) [inline Example/Showcase/Classes/DomainLayer/Chat/ChatService.swift:11]: I don't like the name derivator etc.
DefaultMnemonicDerivationProvider DefaultHDDerivationProvider

**Review by llbartekll — APPROVED** [review]

File: Example/Showcase/Classes/DomainLayer/Chat/ChatService.swift (hunk 5 above)
Comment (flypaper0) [reply Example/Showcase/Classes/DomainLayer/Chat/ChatService.swift:11]: What about `BIP44Provider`?

Comment (llbartekll) [reply Example/Showcase/Classes/DomainLayer/Chat/ChatService.swift:11]: sounds good to me!

**Review by llbartekll — APPROVED** [review]
looks better to me right now!

**Pull request merged**